### PR TITLE
close the connection if we aren't going to retry this OnConnect

### PR DESCRIPTION
**Pull request by mmaxim**

### CODE REVIEW

File: rpc/connection.go [hunk 1]
```diff
@@ -609,6 +609,9 @@ func (c *Connection) doReconnect(ctx context.Context, disconnectStatus Disconnec
 			// A fatal error happened.
 			*reconnectErrPtr = err
 			// short-circuit Retry
+			if err != nil {
+				c.transport.Close()
```
Comment (strib) [inline rpc/connection.go]: Currently the transport is only ever closed on `Shutdown`.  Previously, if `ShouldRetryOnReconnect` returns `false`, this function would return, but I think future calls using the same client would try to reconnect using the same transport.

If the transport is closed here, what will happen to code that relies on the old behavior?  It is required to replace their connection instance? And if the existing code already does that, are they not calling `Shutdown` properly before they create a new connection?

**Comment by mmaxim** [comment]
Added @jzila, whoever would like to look at this.

File: rpc/connection.go (hunk 1 above)
Comment (mmaxim) [inline rpc/connection.go]: Wouldn't existing code just call `Dial` on this transport again? Why would you want to re-use the connection that just failed?

Comment (strib) [inline rpc/connection.go]: Existing code (by that I mean the mdserver/bserver clients) doesn't call `Dial` at all.  It just uses the existing `Connection` object until a ping is lost or it gets a connectivity check request.  From what I understand, the way things work now, a new RPC function call will end up kicking off reconnect again, right?  I'm not sure what happens if you try reconnect on a connection that has a closed transport.

Comment (mmaxim) [inline rpc/connection.go]: I believe `doReconnect` just calls `connect`, which does this:

```
// connect performs the actual connect() and rpc setup.
func (c *Connection) connect(ctx context.Context) error {
	c.log.Debug("Connection: dialing transport")

	// connect
	transport, err := c.transport.Dial(ctx)
	if err != nil {
		c.log.Warning("Connection: error dialing transport: %s", err)
		return err
	}

	client := NewClient(transport, c.errorUnwrapper, c.tagsFunc)
	server := NewServer(transport, c.wef)
```

Comment (mmaxim) [inline rpc/connection.go]: In other words, I think we are calling `Dial` regardless on these reconnect tries, he `Close` I stuck in just makes sure the fd from the previous attempt gets cleaned up. I could be missing something though.

Comment (mmaxim) [inline rpc/connection.go]: Also, this is easy to repro if you just force `OnConnect` to always return an error, and `ShouldRetryOnConnect` to always return false. You should see the process gradually assumulating closed fds.

Comment (strib) [inline rpc/connection.go]: Ah ok thanks.  I didn't realize you could `Dial` again on a closed transport.  So is the real issue that `Dial` doesn't close `t.conn` if it's non-nil, before overwriting it?  Are the other ways to call `Dial` on the same transport twice?

Comment (mmaxim) [inline rpc/connection.go]: Hmm maybe, I guess that could be a better fix. This was the path I found that could do it, there could be others I suppose.

Comment (strib) [inline rpc/connection.go]: Cool.  If you do that, might need some locking in there though, in that case?

Also, still curious about my other question -- how would there ever be 255 `InputCanceledError`s?

Comment (mmaxim) [inline rpc/connection.go]: I don't know about that error, but Gregor returns other errors from there that could pile up. I never actually have seen 255 happen, but I speculated that this issue could be a cause. I've only seen this problem with the Gregor connection, not on kbfsfuse.

**Pull request merged**